### PR TITLE
DR-1079 Performance improvement of batch validation of file id references

### PR DESCRIPTION
**Pull request by ddietterich**
With the usual caveats about doing perf testing on dev, this showed a 33x performance improvement over the naive implementation.
- Added a separate batch size parameter for the validation batch.
- Wrote custom code to validate existence.

### CODE REVIEW

File: src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java [hunk 1]
```diff
@@ -222,15 +222,20 @@ public FireStoreDirectoryEntry retrieveByPath(
     }
 
     public List<String> validateRefIds(
-        Firestore firestore, String collectionId, Collection<String> refIdArray)
+        Firestore firestore, String collectionId, List<String> refIdArray)
         throws InterruptedException {
 
+        int batchSize = configurationService.getParameterValue(FIRESTORE_RETRIEVE_BATCH_SIZE);
+        List<List<String>> batches = ListUtils.partition(refIdArray, batchSize);
```
Comment (fboulnois) [inline src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:229]: Is there any concern essentially doubling the memory use for a large list of refIds?

**Review by fboulnois — APPROVED** [review]
Same question as earlier:

**Review by snf2ye — APPROVED** [review]
Looks good to me!

File: src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java [hunk 2]
```diff
@@ -573,6 +578,35 @@ private void batchStoreDirectoryEntry(
...
+    private List<String> batchValidateIds(
+        Firestore firestore, String collectionId, List<String> batch)
+        throws InterruptedException {
+        CollectionReference datasetCollection = firestore.collection(collectionId);
+
+        List<String> missingIds = new ArrayList<>();
+        List<ApiFuture<QuerySnapshot>> futures = new ArrayList<>();
+        for (String fileId : batch) {
+            Query query = datasetCollection.whereEqualTo("fileId", fileId);
+            futures.add(query.get());
```
Comment (snf2ye) [inline src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:592]: Question: It looks like many of our other query calls are wrapped in a try/catch or are called from a transaction. Is there a reason why we're not doing that here?

Comment (ddietterich) [reply src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:592]: any errors from the query are thrown when we get the results of the query and that happens inside of the try catch.
I could extend the tribe but nothing above there throws ExecutionException.

Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:592]: Gotcha, makes sense!

**Pull request merged**